### PR TITLE
Add queueless dispatch option

### PR DESCRIPTION
**Pull request by kzscisoft**
Adds `PromptDispatcher` for sending metrics and events immediately, this has been kept as an internal option for now as the default should be to not allow oversaturating the server.

Also restructures `simvue.factory` into two components:

* `simvue.factory.proxy`: New home of `Simvue` method for selecting offline/online.
* `simvue.factory.dispatch`: Selection between `QueueDispatcher` and `PromptDispatcher`.

### CODE REVIEW

File: simvue/factory/dispatch/prompt.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        super().__init__(
+            callback=callback,
+            object_types=object_types,
+            termination_trigger=termination_trigger,
+            attributes=attributes,
+        )
+
+    def add_item(self, item: typing.Any, object_type: str) -> None:
+        """Execute callback on the given item"""
+        self._callback([item], object_type, self._attributes)
```
Comment (james-panayis) [inline simvue/factory/dispatch/prompt.py:40]: I think this should record internally how many times `run` has been called, and fail if `add_item` is ever called when `run` hasn't been called exactly once. That way, switching from to the queuing dispatcher should be less likely to cause people new issues

**Comment by james-panayis** [comment]
I prefer "direct" to "prompt", which is less clear.

**Comment by kzscisoft** [comment]
> I prefer "direct" to "prompt", which is less clear.

Changed

File: tests/refactor/test_dispatch.py [hunk 2]
```diff
@@ -137,27 +139,72 @@ def _main(res_queue, index, dispatch_callback=create_callback, term_event=event,
...
+
+    if not append_during_dispatch:
+        for i in range(n_elements):
+            for variable, dispatcher in zip(variables, dispatchers):  
+                dispatcher.add_item({string.ascii_uppercase[i % 26]: i}, variable)
+
+    if append_during_dispatch:
+        for i in range(n_elements):
+            for variable, dispatcher in zip(variables, dispatchers):  
+                dispatcher.add_item({string.ascii_uppercase[i % 26]: i}, variable)
```
Comment (james-panayis) [inline tests/refactor/test_dispatch.py:202]: This seems to be doing the same thing regardles of whether `append_during_dispatch` is `True` or `False`.

Comment (kzscisoft) [reply tests/refactor/test_dispatch.py]: Yep this was a classic copy paste issue

File: tests/refactor/test_dispatch.py [hunk 3]
```diff
@@ -137,27 +139,62 @@ def _main(res_queue, index, dispatch_callback=create_callback, term_event=event,
...
     with pytest.raises(KeyError):
         dispatcher.add_item("blah", "z", False)
 
     trigger.set()
 
-    
+
+@pytest.mark.dispatch
+@pytest.mark.parametrize("multiple", (True, False), ids=("multiple", "single"))
+def test_direct_dispatcher(multiple: booll) -> None:
```
Comment (james-panayis) [inline tests/refactor/test_dispatch.py:166]: booll -> bool

File: simvue/factory/dispatch/queued.py [hunk 4]
```diff
@@ -33,20 +35,20 @@ class Dispatcher(threading.Thread):
     def __init__(
         self,
         callback: typing.Callable[[list[typing.Any], str], None],
-        queue_categories: list[str],
+        object_types: list[str],
         termination_trigger: threading.Event,
         queue_blocking: bool = False,
```
Comment (james-panayis) [inline simvue/factory/dispatch/queued.py:40]: This is never used

File: simvue/factory/dispatch/queued.py [hunk 5]
```diff
@@ -59,26 +61,31 @@ def __init__(
...
         self._callback = callback
-        self._queues = {label: queue.Queue() for label in queue_categories}
+        self._queues = {label: queue.Queue() for label in object_types}
         self._max_read_rate = max_read_rate
         self._max_buffer_size = max_buffer_size
         self._send_timer = 0
-        self._queue_blocking = queue_blocking  # Does nothing
 
-    def add_item(self, item: typing.Any, queue_label: str, blocking: bool) -> None:
+    def add_item(self, item: typing.Any, object_type: str, blocking: bool) -> None:
```
Comment (james-panayis) [inline simvue/factory/dispatch/queued.py:79]: The required "blocking" parameter gives this a different interface to the direct dispatcher, which is not ideal. I think it should be at least defaulted.

**Pull request merged**